File: society_hub/fix_encoding.dart
import 'dart:io';
import 'dart:convert';

void main() async {
  final dir = Directory('lib');
  
  if (!await dir.exists()) {
    print('lib directory not found');
    return;
  }
  
  await for (var entity in dir.list(recursive: true, followLinks: false)) {
    if (entity is File && entity.path.endsWith('.dart')) {
        try {
            // First try reading as UTF-8
            String content = await entity.readAsString(encoding: utf8);
            String newContent = escapeNonAscii(content);
            if (content != newContent) {
                await entity.writeAsString(newContent, encoding: utf8);
                print('Fixed encoding (UTF-8) in: ${entity.path}');
            }
        } catch (e) {
            // If UTF-8 fails
            try {
                // Try reading as latin1 (similar to cp1252)
                String content = await entity.readAsString(encoding: latin1);
                String newContent = escapeNonAscii(content);
                if (content != newContent) {
                    await entity.writeAsString(newContent, encoding: utf8); // write back as utf8
                    print('Fixed encoding (Latin-1 fallback) in: ${entity.path}');
                }
            } catch (e2) {
                print('Failed to read ${entity.path}: $e2');
            }
        }
    }
  }
}

String escapeNonAscii(String text) {
  StringBuffer buffer = StringBuffer();
  for (int i = 0; i < text.runes.length; i++) {
    int codePoint = text.runes.elementAt(i);
    // ASCII is 0-127
    if (codePoint > 127) {
      if (codePoint <= 0xFFFF) {
        String hex = codePoint.toRadixString(16).toUpperCase().padLeft(4, '0');
        buffer.write('\\u$hex');
      } else {
        String hex = codePoint.toRadixString(16).toUpperCase();
        buffer.write('\\u{$hex}');
      }
    } else {
      buffer.writeCharCode(codePoint);
    }
  }
  return buffer.toString();
}

File: society_hub/fix_encoding.py
import os
import re

def escape_non_ascii(text):
    result = []
    for char in text:
        if ord(char) > 127:
            # Check if it fits in 4 hex digits (\uXXXX) or needs 6 (\u{XXXXXX})
            code = ord(char)
            if code <= 0xFFFF:
                result.append(f'\\u{code:04X}')
            else:
                result.append(f'\\u{{{code:X}}}')
        else:
            result.append(char)
    return "".join(result)

def process_directory(directory):
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith('.dart'):
                filepath = os.path.join(root, file)
                try:
                    # Try reading as UTF-8
                    with open(filepath, 'r', encoding='utf-8') as f:
                        content = f.read()
                except UnicodeDecodeError:
                    # If it fails, fallback to cp1252 (Windows)
                    with open(filepath, 'r', encoding='cp1252') as f:
                        content = f.read()
                
                new_content = escape_non_ascii(content)
                
                if new_content != content:
                    with open(filepath, 'w', encoding='utf-8') as f:
                        f.write(new_content)
                    print(f"Fixed encoding in: {filepath}")

if __name__ == "__main__":
    process_directory("f:/society app/society_hub/lib")
